tests/utils_/test_parse_uri.py

```diff
@@ -0,0 +1,90 @@
+from unittest.mock import MagicMock, patch
+
+import pymongo
+from django.test import SimpleTestCase
+
+from django_mongodb import parse_uri
+
+URI = "mongodb+srv://myDatabaseUser:D1fficultP%40ssw0rd@cluster0.example.mongodb.net/myDatabase?retryWrites=true&w=majority&tls=false"
+
+
+class ParseURITests(SimpleTestCase):
+    def setUp(self):
+        self.srv_record = MagicMock()
+        self.srv_record.target.to_text.return_value = "cluster0.example.mongodb.net"
+        self.patcher = patch("dns.resolver.resolve", return_value=[self.srv_record])
+        self.mock_resolver = self.patcher.start()
+        self.addCleanup(self.patcher.stop)
+
+    @patch("dns.resolver.resolve")
+    def test_simple_uri(self, mock_resolver):
+        settings_dict = parse_uri("mongodb://cluster0.example.mongodb.net/myDatabase")
+        self.assertEqual(settings_dict["ENGINE"], "django_mongodb")
+        self.assertEqual(settings_dict["NAME"], "myDatabase")
+        self.assertEqual(settings_dict["HOST"], "cluster0.example.mongodb.net")
+
+    @patch("dns.resolver.resolve")
+    def test_no_database(self, mock_resolver):
+        settings_dict = parse_uri("mongodb://cluster0.example.mongodb.net/")
+        self.assertEqual(settings_dict["ENGINE"], "django_mongodb")
+        self.assertIsNone(settings_dict["NAME"])
```

Comment:
Turns out, based on the spec for `NAME` this should end up returning an empty string if it comes back as `None`.
https://docs.djangoproject.com/en/5.1/ref/settings/#name

Reply:
Additionally, in this case, I think it's best to still have the `name` as a passed through kwarg because Atlas URIs do often not provide a database name. I'll provide an example in the code above about how this would work.

Reply:
If the URI does not contain a database name, then we don't need to add `NAME`. 1519e0f484682f2ad7b89bbb670c0c9c9fac0498

Reply:
I don't think the None/'' distinction matters. What does it mean to use Django + MongoDB without specifying a database name? I think NAME must be required, although it's not the job of `parse_uri()` to raise an error. Currently, when trying runserver without NAME:
```
Traceback (most recent call last):
  File "/opt/python3.12.0/lib/python3.12/threading.py", line 1052, in _bootstrap_inner
    self.run()
  File "/opt/python3.12.0/lib/python3.12/threading.py", line 989, in run
    self._target(*self._args, **self._kwargs)
  File "/home/tim/code/django/django/utils/autoreload.py", line 64, in wrapper
    fn(*args, **kwargs)
  File "/home/tim/code/django/django/core/management/commands/runserver.py", line 136, in inner_run
    self.check_migrations()
  File "/home/tim/code/django/django/core/management/base.py", line 581, in check_migrations
    executor = MigrationExecutor(connections[DEFAULT_DB_ALIAS])
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/db/migrations/executor.py", line 18, in __init__
    self.loader = MigrationLoader(self.connection)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/db/migrations/loader.py", line 58, in __init__
    self.build_graph()
  File "/home/tim/code/django/django/db/migrations/loader.py", line 235, in build_graph
    self.applied_migrations = recorder.applied_migrations()
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/db/migrations/recorder.py", line 89, in applied_migrations
    if self.has_table():
       ^^^^^^^^^^^^^^^^
  File "/home/tim/code/django/django/db/migrations/recorder.py", line 64, in has_table
    tables = self.connection.introspection.table_names(cursor)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/introspection.py", line 10, in table_names
    return sorted([x["name"] for x in self.connection.database.list_collections()])
                                      ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/base.py", line 150, in __getattr__
    return getattr(self, attr)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/base.py", line 150, in __getattr__
    return getattr(self, attr)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/tim/code/django-mongodb/django_mongodb/base.py", line 150, in __getattr__
    return getattr(self, attr)
           ^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 741 more times]
RecursionError: maximum recursion depth exceeded
```

Reply:
That's the bug I found a while back! OK I put it back 56fe609393d8cde55b466f1f6429ce4d12460566

Reply:
Jib said, "I think it's best to still have the name as a passed through kwarg because Atlas URIs do often not provide a database name." By reading the documentation, I hope the user should be able to understand they can append the database name to their URL. Offering a separate keyword increases complexity and makes it possible for them to provide two conflicting values.

Reply:
`NAME` has to be set in `DATABASES` else Django crashes. Once #196 is merged that will be enforced by `get_connect_params`.  So end users that try to do `parse_uri("mongodb://localhost")` (or something like it, without setting `NAME`) will get an error from `get_connect_params`.

Our `parse_uri` is inspired by `dj-database-url` which is inspired by [The Twelve-Factor App](https://12factor.net/backing-services) which prioritizes what @timgraham was arguing about with `HOST`. `HOST` should contain a `HOST` and nothing else so that:

Via  [Twelve-Factor App](https://12factor.net/backing-services)

> Resources can be attached to and detached from deploys at will. For example, if the app’s database is misbehaving due to a hardware issue, the app’s administrator might spin up a new database server restored from a recent backup. The current production database could be detached, and the new database attached – all without any code changes.

`dj-database-url` set the precedent that additional `kwargs` passed into `parse_uri` are passed to the `DATABASES` dictionary and `NAME` has already been parsed in the URL, so we definitely would not consider adding an additional `kwarg` for `NAME`. As @timgraham says, just direct users to the documentation that says "The database `NAME` is part of the URL" per  [Twelve-Factor App](https://12factor.net/backing-services).

Reply:
Reading https://www.mongodb.com/docs/manual/reference/connection-string/, one thing that could be confusing is that the example connection URL is: `mongodb+srv://[username:password@]host[/[defaultauthdb][?options]]`.

Does it imply that the authentication database could be different from the database that the user may want to store their Django models in? Reading https://pymongo.readthedocs.io/en/stable/examples/authentication.html#default-database-and-authsource, it looks like `defaultauthdb` can actually be the default database and you can specify a different authentication db using `authSource` in the query string. So I think our implementation is okay.

Reply:
Right, I think it's OK too, but playing devil's advocate for @Jibola , Django says `NAME` [can be an empty string](https://docs.djangoproject.com/en/5.1/ref/settings/#name) and if we assume that MongoDB can auto-assign a collection name then we're back to the logistics of supporting an empty string `NAME` which involves an alternative fix for #196 and additional fixes for `parse_uri`. I don't have a strong preference either way and may even lean towards supporting the empty string `NAME` but I'll pass back to @Jibola for additional feedback.

I will note though that even though Django says `NAME` can be an empty string, that appears just a practical matter for Django to leave it to the back end to decide. If your database is SQLite and your `NAME` is empty what happens? I think SQLite picks a file name. If your database is Postgresql and your `NAME` is empty what happens? I think you get an error, but I'm not sure. Same with MySQL. If the MongoDB preference is for `NAME` to be an empty string and there is no error, then I imagine we can get Django to play along with that.

Reply:
Django's documentation says the the default of NAME is an empty string, but that doesn't mean it's a valid configuration. Some database backends like sqlite3 [raise an error if NAME isn't provided](https://github.com/django/django/blob/5e998d717f7b4220a1728bd49b66ca0162e2a6cb/django/db/backends/sqlite3/base.py#L157-L161), others like have the notion of [a "no db" connection](https://github.com/django/django/blob/5e998d717f7b4220a1728bd49b66ca0162e2a6cb/django/db/backends/postgresql/base.py#L240-L264) (`'NAME': None`) that is of some limited use. For example, when testing on PostgreSQL, "Normally Django will use a connection to the 'postgres' database to avoid running initialization queries against the production."

If there are cases where using Django + MongoDB without a database name is a valid configuration, we need the details. The behavior I see is:
```
>>> db_name = ''
>>> MongoClient(...)[db_name]
...
  File ".../pymongo/synchronous/mongo_client.py", line 1183, in __getitem__
    return database.Database(self, name)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "../pymongo/synchronous/database.py", line 133, in __init__
    _check_name(name)
  File ".../pymongo/database_shared.py", line 27, in _check_name
    raise InvalidName("database name cannot be the empty string")
pymongo.errors.InvalidName: database name cannot be the empty string
```

Reply:
Well played